### PR TITLE
Implement event sending part of /rooms/$roomid/send/$type/$txnid

### PR DESCRIPTION
**Pull request by Kegsay**
This involves:
 - Parsing the HTTP request
 - Requesting auth events from the roomserver via the Query API
 - Building the event
 - Doing auth checks on the event
 - Sending it to the roomserver input log

Tested this with a roomserver instance and it all seems to work as expected. There's an increasing amount of duplication and pain points I'm seeing, but will fix those in subsequent PRs.

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go [hunk 1]
```diff
@@ -3,22 +3,150 @@ package writers
...
+		return httputil.LogThenError(req, err)
+	}
+
+	return util.JSONResponse{
+		Code: 200,
+		JSON: sendMessageResponse{e.EventID()},
+	}
+}
+
+func sendToRoomserver(e gomatrixserverlib.Event, producer sarama.SyncProducer, topic string) error {
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: This is presumably slightly duplicated in createRoom?

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go [hunk 2]
```diff
@@ -3,22 +3,150 @@ package writers
...
+		Kind:         api.KindNew,
+		Event:        e.JSON(),
+		AuthEventIDs: authEventIDs,
+	}
+
+	value, err := json.Marshal(ire)
+	if err != nil {
+		return err
+	}
+	var m sarama.ProducerMessage
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: It might be nice if the kafka specific bits were hidden behind an interface that took an InputRoomEvent.

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go [hunk 3]
```diff
@@ -3,22 +3,150 @@ package writers
...
+	m.Topic = topic
+	m.Key = sarama.StringEncoder(e.EventID())
+	m.Value = sarama.ByteEncoder(value)
+	if _, _, err := producer.SendMessage(&m); err != nil {
+		return err
+	}
+	return nil
+}
+
+func stateNeeded(builder *gomatrixserverlib.EventBuilder) (requiredStateEvents []common.StateKeyTuple, err error) {
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: Much of this logic seems duplicated in `authEventsFromStateNeeded`. Would it make sense to make authEventsFromStateNeeded use this function when working out which common.StateKeyTuple are needed?

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 1 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: Yes, albeit this is the single event form and the other is the bulk form. I will factor it out in a separate PR.

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 2 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: Instead of passing in a `sarama.Producer`, I'm happy to pass in an interface which will send `InputRoomEvents` to the roomserver. I'd like to do that in another PR though.

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 3 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: Yes it would make sense. I would like to do that in a separate PR though.

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 1 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: https://github.com/matrix-org/dendrite/pull/40

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 2 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: https://github.com/matrix-org/dendrite/pull/40

File: src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go (hunk 3 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/writers/sendmessage.go]: https://github.com/matrix-org/dendrite/pull/44

**Comment by Kegsay** [comment]
Changes have now happened.
